contributors/design-proposals/storage/container-storage-interface-cluster-registry.md

```diff
@@ -0,0 +1,156 @@
+
+### Links
+
+* [Container Storage Interface (CSI) Spec](https://github.com/container-storage-interface/spec/blob/master/spec.md)
+* [CSI Volume Plugins in Kubernetes Design Doc](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md)
+
+## Objective
+
+The objective of this document is to document all the requirements for enabling a cluster level registry for a CSI compliant volume drivers in Kubernetes.
+
+## Goals
+
+* Allow a CSI volume driver (that opts-in to new mechanism) to configure how Kubernetes should interact with it.
+* Improve discoverability by users/cluster admins of a CSI volume driver (that opts-in to new mechanism) deployed on the Kubernetes cluster.
+
+## Non-Goals
+
+* Define how CSI volume driver should be deployed on Kubernetes.
+* Require use of new CSI cluster registry mechanism for all CSI volume drivers deployed on Kubernetes (the mechanism will be optional/opt-in).
+
+## Design Overview
+
+A new custom resource definition will automatically be installed on Kubernetes clusters.
+
+Upon deployment a driver MAY create a new custom resource object.
+
+## Design Details
+
+### `CSIDriver` Object
+
+#### Proposed API
+
+```go
+// CSIDriver captures information about a Container Storage Interface (CSI)
+// volume driver deployed on the cluster.
+//
+// CSIDriver objects are non-namespaced.
+type CSIDriver struct {
+    metav1.TypeMeta   `json:",inline"`
+
+    // Standard object metadata.
+    // More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+    metav1.ObjectMeta `json:"metadata"`
+
+    // Specification describing the CSI volume driver and any custom
+    // configuration for it.
+    Spec CSIDriverSpec `json:"spec"`
+}
+
+// CSIDriverList is a collection of CSIDriver objects.
+type CSIDriverList struct {
+    metav1.TypeMeta `json:",inline"`
+    
+    // Standard list metadata
+    // More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+    metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+    // Items is the list of CSIDrivers
+    Items []CSIDriver `json:"items"`
+}
+
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    // Driver indicates the name of the CSI driver that this object refers to.
+    // This MUST be the same name returned by the CSI GetPluginName() call for
+    // that driver.
+    Driver string `json:"driver"`
+    
+    // Indicates this CSI volume driver requires an attach operation (because it
+    // implements the CSI ControllerPublishVolume() method), and that Kubernetes
+    // should call attach and wait for any attach operation to complete before
+    // proceeding to mounting.
+    // If value is not specified, default is true -- meaning attach will be
+    // called.
+    // +optional
+    AttachRequired *bool `json:"attachRequired"`
```

Comment:
Probably a dump question. 

Isn't this value already exposed from `ControllerServiceCapability` as `PUBLISH_UNPUBLISH_VOLUME`? Can't kubelet make decision based on that value?

Reply:
The problem is where the call is being made. Kubelet is not doing the attach. A pod floating in the cluster takes attachment requests, converts them into the driver calls for attach/detach.

You could have the attacher sidecar running, probe the driver as you say, and then determine that it doesn't ever have to do anything. That does work today.

But that's a fair amount of extra pluming running in the cluster that isn't needed.

The flag above tells Kubernetes to skip entirely any of the handshaking at the cluster level required to do that attach so the sidecar that does nothing and its driver doesn't even need to exist.